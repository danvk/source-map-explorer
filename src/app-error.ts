import { formatPercent } from './helpers';
import { ErrorCode } from './index';

// If we need advanced error consider using https://github.com/joyent/node-verror
export class AppError extends Error {
  code?: ErrorCode;
  cause?: Error;

  constructor(errorContext: ErrorContext, error?: NodeJS.ErrnoException) {
    super();
    // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
    Object.setPrototypeOf(this, AppError.prototype);

    const message = getErrorMessage(errorContext);

    this.message = error ? `${message}: ${error.message}` : message;
    this.code = errorContext.code;

    Error.captureStackTrace(this, AppError);
  }
}

export const SOURCE_MAP_INFO_URL =
  'https://github.com/danvk/source-map-explorer/blob/master/README.md#generating-source-maps';

interface CommonErrorContext {
  code: 'NoBundles' | 'NoSourceMap' | 'CannotSaveFile' | 'CannotCreateTempFile' | 'Unknown';
}

interface OneSourceSourceMapErrorContext {
  code: 'OneSourceSourceMap';
  filename: string;
}

interface UnmappedBytesErrorContext {
  code: 'UnmappedBytes';
  totalBytes: number;
  unmappedBytes: number;
}

interface InvalidMappingLineErrorContext {
  code: 'InvalidMappingLine';
  generatedLine: number;
  maxLine: number;
}

interface InvalidMappingColumnErrorContext {
  code: 'InvalidMappingColumn';
  generatedLine: number;
  generatedColumn: number;
  maxColumn: number;
}

interface CannotOpenTempFileErrorContext {
  code: 'CannotOpenTempFile';
  error: Buffer;
  tempFile: string;
}

export type ErrorContext =
  | CommonErrorContext
  | OneSourceSourceMapErrorContext
  | UnmappedBytesErrorContext
  | InvalidMappingLineErrorContext
  | InvalidMappingColumnErrorContext
  | CannotOpenTempFileErrorContext;

export function getErrorMessage(context: ErrorContext): string {
  switch (context.code) {
    case 'NoBundles':
      return 'No file(s) provided';

    case 'NoSourceMap':
      return `Unable to find a source map.
See ${SOURCE_MAP_INFO_URL}`;

    case 'OneSourceSourceMap': {
      return `Your source map only contains one source (${context.filename}).
This can happen if you use browserify+uglifyjs, for example, and don't set the --in-source-map flag to uglify.
See ${SOURCE_MAP_INFO_URL}`;
    }

    case 'UnmappedBytes': {
      const { unmappedBytes, totalBytes } = context;

      const bytesString = formatPercent(unmappedBytes, totalBytes, 2);

      return `Unable to map ${unmappedBytes}/${totalBytes} bytes (${bytesString}%)`;
    }

    case 'InvalidMappingLine': {
      const { generatedLine, maxLine } = context;

      return `Your source map refers to generated line ${generatedLine}, but the source only contains ${maxLine} line(s).
Check that you are using the correct source map.`;
    }

    case 'InvalidMappingColumn': {
      const { generatedLine, generatedColumn, maxColumn } = context;

      return `Your source map refers to generated column ${generatedColumn} on line ${generatedLine}, but the source only contains ${maxColumn} column(s) on that line.
Check that you are using the correct source map.`;
    }

    case 'CannotSaveFile':
      return 'Unable to save HTML to file';

    case 'CannotCreateTempFile':
      return 'Unable to create a temporary HTML file';

    case 'CannotOpenTempFile': {
      const { error, tempFile } = context;

      return `Unable to open web browser. ${error.toString().trim()}
Either run with --html, --json, --tsv, --file, or view HTML for the visualization at:
${tempFile}`;
    }

    default:
      return 'Unknown error';
  }
}
