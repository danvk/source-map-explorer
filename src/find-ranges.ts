export function findCoveredBytes(coveredRanges, moduleRanges): { [module: string]: number } {
  const sortedCoverageRanges = coveredRanges.sort((a, b) => a.start - b.start);
  const sortedModuleRanges = moduleRanges.sort((a, b) => a.start - b.start);

  let i = 0;
  let j = 0;
  const sizes = {};

  while (i < moduleRanges.length && j < coveredRanges.length) {
    const moduleRange = sortedModuleRanges[i];
    const coverageRange = sortedCoverageRanges[j];

    if (moduleRange.start <= coverageRange.end && moduleRange.end >= coverageRange.start) {
      // overlaps, calculate amount, move to next coverage range
      const end = Math.min(coverageRange.end, moduleRange.end);
      const start = Math.max(moduleRange.start, coverageRange.start);
      if (sizes[moduleRange.module] === undefined) {
        sizes[moduleRange.module] = 0;
      }
      sizes[moduleRange.module] += end - start + 1;

      if (
        sortedModuleRanges[i + 1] !== undefined &&
        sortedModuleRanges[i + 1].start <= coverageRange.end &&
        sortedModuleRanges[i + 1].end >= coverageRange.start
      ) {
        // next module also overlaps current coverage range, advance to next module instead of advancing coverage
        i++;
      } else {
        // check next coverage range, it may also overlap this module range
        j++;
      }
    } else if (moduleRange.end < coverageRange.start) {
      // module comes entirely before coverageRange, check next module range
      i++;
    }
    if (coverageRange.end < moduleRange.start) {
      // module range comes entirely after coverage range, check next coverage range
      j++;
    }
  }

  return sizes;
}
