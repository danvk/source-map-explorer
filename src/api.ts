import glob from 'glob';
import { partition, flatMap, isString } from 'lodash';

import { exploreBundle, UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY } from './explore';
import { AppError, getErrorMessage } from './app-error';
import {
  BundlesAndFileTokens,
  ExploreOptions,
  ExploreResult,
  Bundle,
  ExploreErrorResult,
  ExploreBundleResult,
} from './index';
import { formatOutput, saveOutputToFile } from './output';

/**
 * Analyze bundle(s)
 */
export async function explore(
  bundlesAndFileTokens: BundlesAndFileTokens,
  options: ExploreOptions = {}
): Promise<ExploreResult> {
  bundlesAndFileTokens = Array.isArray(bundlesAndFileTokens)
    ? bundlesAndFileTokens
    : [bundlesAndFileTokens];

  if (bundlesAndFileTokens.length === 0) {
    throw new AppError({ code: 'NoBundles' });
  }

  // Separate bundles from file tokens
  const [fileTokens, bundles] = partition(bundlesAndFileTokens, isString);

  // Get bundles from file tokens
  bundles.push(...getBundles(fileTokens));

  const results = await Promise.all(
    bundles.map(bundle =>
      exploreBundle(bundle, options).catch<ExploreErrorResult>(error =>
        onExploreError(bundle, error)
      )
    )
  );

  const exploreResult = getExploreResult(results, options);

  // Reject if none of results is successful
  if (exploreResult.bundles.length === 0) {
    return Promise.reject(exploreResult);
  }

  saveOutputToFile(exploreResult, options);

  return exploreResult;
}

/**
 * Expand list of file tokens into a list of bundles
 */
export function getBundles(fileTokens: string[]): Bundle[] {
  const filenames = flatMap(fileTokens, filePath =>
    glob.hasMagic(filePath) ? expandGlob(filePath) : filePath
  );

  const [mapFilenames, codeFilenames] = partition(filenames, filename => filename.endsWith('.map'));

  return codeFilenames.map<Bundle>(code => ({
    code,
    map: mapFilenames.find(filename => filename === `${code}.map`),
  }));
}

function expandGlob(pattern: string): string[] {
  // Make sure pattern match `.map` files as well
  if (pattern.endsWith('.js')) {
    pattern = `${pattern}?(.map)`;
  }

  return glob.sync(pattern);
}

export function getBundleName(bundle: Bundle): string {
  return Buffer.isBuffer(bundle.code) ? 'Buffer' : bundle.code;
}

/**
 * Handle error during bundle processing
 */
function onExploreError(bundle: Bundle, error: NodeJS.ErrnoException): ExploreErrorResult {
  return {
    bundleName: getBundleName(bundle),
    code: error.code || 'Unknown',
    message: error.message,
    error,
  };
}

function getExploreResult(
  results: (ExploreBundleResult | ExploreErrorResult)[],
  options: ExploreOptions
): ExploreResult {
  const [bundles, errors] = partition(
    results,
    (result): result is ExploreBundleResult => 'files' in result
  );

  errors.push(...getPostExploreErrors(bundles));

  return {
    bundles,
    errors,
    ...(bundles.length > 0 && { output: formatOutput(bundles, options) }),
  };
}

const SPECIAL_FILENAMES = [UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY];

function getPostExploreErrors(exploreBundleResults: ExploreBundleResult[]): ExploreErrorResult[] {
  const errors: ExploreErrorResult[] = [];

  const isSingleBundle = exploreBundleResults.length === 1;

  for (const result of exploreBundleResults) {
    const { bundleName, files, totalBytes } = result;

    // Check if source map contains only one file - this make result useless when exploring single bundle
    if (isSingleBundle) {
      const filenames = Object.keys(files).filter(
        filename => !SPECIAL_FILENAMES.includes(filename)
      );
      if (filenames.length === 1) {
        errors.push({
          bundleName,
          isWarning: true,
          code: 'OneSourceSourceMap',
          message: getErrorMessage({ code: 'OneSourceSourceMap', filename: filenames[0] }),
        });
      }
    }

    const unmappedBytes = files[UNMAPPED_KEY];
    if (unmappedBytes) {
      errors.push({
        bundleName,
        isWarning: true,
        code: 'UnmappedBytes',
        message: getErrorMessage({ code: 'UnmappedBytes', unmappedBytes, totalBytes }),
      });
    }
  }

  return errors;
}
