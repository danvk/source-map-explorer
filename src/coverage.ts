import { Bundle, Coverage, ColumnsRange, MappingRange, FileDataMap } from './index';
import { getFileContent, detectEOL } from './helpers';
import { AppError } from './app-error';

/**
 * Convert one-line coverage ranges (exclusive) into per line ranges (inclusive)
 */
function convertRangesToLinesRanges(coverage: Coverage): ColumnsRange[][] {
  const { ranges, text } = coverage;

  const eol = detectEOL(text);
  const eolLength = eol.length;
  const lines = text.split(eol);

  // Current line offset
  let offset = 0;

  const lineRanges = lines.map(line => {
    const lineLength = line.length;

    if (lineLength === 0) {
      return [];
    }

    // Exclusive line start/end
    const lineStart = offset;
    const lineEnd = offset + lineLength;

    const lineRanges = ranges.reduce<ColumnsRange[]>((result, { start, end }) => {
      // Inclusive range start/end within the line
      const startIndex = start - lineStart;
      const endIndex = end - lineStart - 1;
      const lineEndIndex = lineLength - 1;

      if (start <= lineStart && lineEnd <= end) {
        // Range includes line range
        result.push({ start: 0, end: lineEndIndex });
      } else if (lineStart <= start && end <= lineEnd) {
        // Line range includes range
        result.push({ start: startIndex, end: endIndex });
      } else if (lineStart <= start && start <= lineEnd) {
        // Range starts within line range
        result.push({ start: startIndex, end: lineEndIndex });
      } else if (lineStart <= end && end <= lineEnd) {
        // Range ends within line range
        result.push({ start: 0, end: endIndex });
      }

      return result;
    }, []);

    // Move to next line jumping over EOL character
    offset = lineEnd + eolLength;

    return lineRanges;
  });

  return lineRanges;
}

const PATH_SEPARATOR_REGEX = /[/\\]/;

function getPathParts(path: string): string[] {
  return path.split(PATH_SEPARATOR_REGEX).filter(Boolean);
}

/**
 * Match coverages' ranges to bundles by comparing coverage URL and bundle filename
 */
export function addCoverageRanges(bundles: Bundle[], coverageFilename?: string): Bundle[] {
  if (!coverageFilename) {
    return bundles;
  }

  try {
    const coverages: Coverage[] = JSON.parse(getFileContent(coverageFilename));

    const coveragePaths = coverages.map(({ url }) =>
      getPathParts(new URL(url).pathname || '').reverse()
    );

    const bundlesPaths = bundles.reduce<[string[], number][]>((result, { code }, index) => {
      if (!Buffer.isBuffer(code)) {
        result.push([getPathParts(code).reverse(), index]);
      }

      return result;
    }, []);

    coveragePaths.forEach((partsA, coverageIndex) => {
      let matchingBundles = [...bundlesPaths];

      // Start from filename and go up to path root
      for (let i = 0; i < partsA.length; i++) {
        matchingBundles = matchingBundles.filter(
          ([partsB]) => i < partsB.length && partsB[i] === partsA[i]
        );

        // Stop when exact (among bundles) match found or no matches found
        if (matchingBundles.length <= 1) {
          break;
        }
      }

      if (matchingBundles.length === 1) {
        const [[, bundleIndex]] = matchingBundles;

        bundles[bundleIndex].coverageRanges = convertRangesToLinesRanges(coverages[coverageIndex]);
      }
    });
  } catch (error) {
    throw new AppError({ code: 'CannotOpenCoverageFile' }, error);
  }

  if (bundles.every(({ coverageRanges }) => coverageRanges === undefined)) {
    throw new AppError({ code: 'NoCoverageMatches' });
  }

  return bundles;
}

/**
 * Find overlaps in arrays of column ranges, using ratcheting pointers instead of nested loops for
 * O(n) runtime instead of O(n^2)
 */
function findCoveredMappingRanges(
  mappingRanges: MappingRange[],
  coveredRanges: ColumnsRange[]
): MappingRange[] {
  let i = 0;
  let j = 0;

  const result: MappingRange[] = [];

  while (i < mappingRanges.length && j < coveredRanges.length) {
    const mappingRange = mappingRanges[i];
    const coveredRange = coveredRanges[j];

    if (mappingRange.start <= coveredRange.end && coveredRange.start <= mappingRange.end) {
      // Overlaps, calculate amount, move to next coverage range
      const end = Math.min(coveredRange.end, mappingRange.end);
      const start = Math.max(mappingRange.start, coveredRange.start);

      result.push({
        start,
        end,
        source: mappingRange.source,
      });

      if (
        mappingRanges[i + 1] !== undefined &&
        mappingRanges[i + 1].start <= coveredRange.end &&
        mappingRanges[i + 1].end >= coveredRange.start
      ) {
        // Next module also overlaps current coverage range, advance to next module instead of advancing coverage
        i++;
      } else {
        // Check next coverage range, it may also overlap this module range
        j++;
      }
    } else if (mappingRange.end < coveredRange.start) {
      // Module comes entirely before coverageRange, check next module range
      i++;
    }
    if (coveredRange.end < mappingRange.start) {
      // Module range comes entirely after coverage range, check next coverage range
      j++;
    }
  }

  return result;
}

/**
 * Set covered size for files
 */
export function setCoveredSizes(
  line: string,
  files: FileDataMap,
  mappingRanges: MappingRange[],
  coveredRanges: ColumnsRange[]
): FileDataMap {
  findCoveredMappingRanges(mappingRanges, coveredRanges).forEach(({ start, end, source }) => {
    const rangeByteLength = Buffer.byteLength(line.substring(start, end + 1));

    let coveredSize = files[source].coveredSize || 0;

    coveredSize += rangeByteLength;

    files[source].coveredSize = coveredSize;
  });

  return files;
}

const percentColors = [
  { percent: 0.0, color: { r: 0xff, g: 0x00, b: 0 } },
  { percent: 0.5, color: { r: 0xff, g: 0xff, b: 0 } },
  { percent: 1.0, color: { r: 0x00, g: 0xff, b: 0 } },
];

/**
 * Get heat map color by coverage percent
 */
export function getColorByPercent(percent: number): string {
  let i = 1;

  for (; i < percentColors.length - 1; i++) {
    if (percent < percentColors[i].percent) {
      break;
    }
  }

  const lowerColor = percentColors[i - 1];
  const upperColor = percentColors[i];
  const rangeWithinColors = upperColor.percent - lowerColor.percent;
  const rangePercent = (percent - lowerColor.percent) / rangeWithinColors;
  const percentLower = 1 - rangePercent;
  const percentUpper = rangePercent;

  const r = Math.floor(lowerColor.color.r * percentLower + upperColor.color.r * percentUpper);
  const g = Math.floor(lowerColor.color.g * percentLower + upperColor.color.g * percentUpper);
  const b = Math.floor(lowerColor.color.b * percentLower + upperColor.color.b * percentUpper);

  return `rgb(${r}, ${g}, ${b})`;
}
