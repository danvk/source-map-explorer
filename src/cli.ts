#!/usr/bin/env node

import yargs from 'yargs';
import fs from 'fs';
import temp from 'temp';
import open from 'open';
import chalk from 'chalk';
import { groupBy, isString } from 'lodash';

import { explore } from './api';
import { ExploreOptions, ReplaceMap, ExploreResult } from './index';
import { AppError, getErrorMessage, ErrorContext } from './app-error';

/** Parsed CLI arguments */
interface Arguments {
  _: string[];
  json?: string;
  tsv?: string;
  html?: string;
  onlyMapped?: boolean;
  excludeSourceMap?: boolean;
  noRoot?: boolean;
  replace?: string[];
  with?: string[];
}

function parseArguments(): Arguments {
  const argv = yargs
    .strict()
    .scriptName('source-map-explorer')
    .usage('Analyze and debug space usage through source maps.')
    .usage('Usage:')
    .usage(
      '$0 script.js [script.js.map] [--json [result.json] | --html [result.html] | --tsv [result.csv]] [-m | --only-mapped] [--exclude-source-map] [--replace=BEFORE_1 BEFORE_2 --with=AFTER_1 AFTER_2] [--no-root] [--version] [--help | -h]'
    )
    .example('$0 script.js script.js.map', 'Explore bundle')
    .example('$0 script.js', 'Explore bundle with inline source map')
    .example('$0 dist/js/*.*', 'Explore all bundles inside dist/js folder')
    .example('$0 script.js --tsv', 'Explore and output result as TSV to stdout')
    .example('$0 script.js --json result.json', 'Explore and save result as JSON to the file')
    .demandCommand(1, 'At least one js file must be specified')
    .options({
      json: {
        type: 'string',
        description:
          'If filename specified save output as JSON to specified file otherwise output to stdout.',
        conflicts: ['tsv', 'html'],
      },
      tsv: {
        type: 'string',
        description:
          'If filename specified save output as TSV to specified file otherwise output to stdout.',
        conflicts: ['json', 'html'],
      },
      html: {
        type: 'string',
        description:
          'If filename specified save output as HTML to specified file otherwise output to stdout rather than opening a browser.',
        conflicts: ['json', 'tsv'],
      },

      'only-mapped': {
        alias: 'm',
        type: 'boolean',
        description:
          'Exclude "unmapped" bytes from the output. This will result in total counts less than the file size',
      },

      'exclude-source-map': {
        type: 'boolean',
        description: 'Exclude source map comment size from output',
      },

      'no-root': {
        type: 'boolean',
        description:
          'To simplify the visualization, source-map-explorer will remove any prefix shared by all sources. If you wish to disable this behavior, set --no-root.',
      },

      replace: {
        type: 'string',
        array: true,
        description:
          'Apply a simple find/replace on source file names. This can be used to fix some oddities with paths which appear in the source map generation process. Accepts regular expressions.',
        implies: 'with',
      },
      with: {
        type: 'string',
        array: true,
        description: 'See --replace.',
        implies: 'replace',
      },
    })
    .group(['json', 'tsv', 'html'], 'Output:')
    .group(['replace', 'with'], 'Replace:')
    .help('h')
    .alias('h', 'help')
    .showHelpOnFail(false, 'Specify --help for available options')
    .wrap(null) // Do not limit line length
    .parserConfiguration({
      'boolean-negation': false, // Allow --no-root
    })
    .check(argv => {
      if (argv.replace && argv.with && argv.replace.length !== argv.with.length) {
        throw new Error('--replace flags must be paired with --with flags');
      }

      return true;
    })
    .parse();

  // Trim extra quotes
  const quoteRegex = /'/g;
  argv._ = argv._.map(path => path.replace(quoteRegex, ''));

  return argv;
}

export function logError(message: string | ErrorContext, error?: Error): void {
  if (!isString(message)) {
    message = getErrorMessage(message);
  }

  if (error) {
    console.error(chalk.red(message), error);
  } else {
    console.error(chalk.red(message));
  }
}

export function logWarn(message: string): void {
  console.warn(chalk.yellow(message));
}

export function logInfo(message: string): void {
  console.log(chalk.green(message));
}

/**
 * Create options object for `explore` method
 */
function getExploreOptions(argv: Arguments): ExploreOptions {
  let replaceMap: ReplaceMap | undefined;
  const replaceItems = argv.replace;
  const withItems = argv.with;

  if (replaceItems && withItems) {
    replaceMap = replaceItems.reduce<ReplaceMap>((result, item, index) => {
      result[item] = withItems[index];

      return result;
    }, {});
  }

  return {
    output: {
      // By default CLI needs result in HTML in order to create a temporary file
      format: isString(argv.json) ? 'json' : isString(argv.tsv) ? 'tsv' : 'html',
      filename: argv.json || argv.tsv || argv.html,
    },
    replaceMap,
    onlyMapped: argv.onlyMapped,
    excludeSourceMapComment: argv.excludeSourceMap,
    noRoot: argv.noRoot,
  };
}

/**
 * Write HTML content to a temporary file and open the file in a browser
 */
async function writeHtmlToTempFile(html?: string): Promise<void> {
  if (!html) {
    return;
  }

  try {
    const tempFile = temp.path({ prefix: 'sme-result-', suffix: '.html' });
    fs.writeFileSync(tempFile, html);

    const childProcess = await open(tempFile);

    if (childProcess.stderr) {
      // Catch error output from child process
      childProcess.stderr.once('data', (error: Buffer) => {
        logError({ code: 'CannotOpenTempFile', tempFile, error });
      });
    }
  } catch (error) {
    throw new AppError({ code: 'CannotCreateTempFile' }, error);
  }
}

function outputErrors({ errors }: ExploreResult): void {
  if (errors.length === 0) {
    return;
  }

  // Group errors by bundle name
  const groupedErrors = groupBy(errors, 'bundleName');

  Object.entries(groupedErrors).forEach(([bundleName, errors]) => {
    console.group(bundleName);

    const hasManyErrors = errors.length > 1;
    errors.forEach((error, index) => {
      const message = `${hasManyErrors ? `${index + 1}. ` : ''}${error.message}`;

      if (error.isWarning) {
        logWarn(message);
      } else {
        logError(message);
      }
    });

    console.groupEnd();
  });
}

if (require.main === module) {
  const argv = parseArguments();

  const isOutputFormatSpecified = [argv.json, argv.tsv, argv.html].some(isString);

  const options = getExploreOptions(argv);

  explore(argv._, options)
    .then(result => {
      if (isOutputFormatSpecified && options.output) {
        const filename = options.output.filename;

        if (filename) {
          logInfo(`Output saved to ${filename}`);
          outputErrors(result);
        } else {
          console.log(result.output);
        }
      } else {
        writeHtmlToTempFile(result.output).then(() => {
          outputErrors(result);
        });
      }
    })
    .catch(error => {
      if (error.errors) {
        outputErrors(error);
      } else {
        logError('Failed to explore', error);
      }
    });
}
