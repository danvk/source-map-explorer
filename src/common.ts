import path from 'path';
import fs from 'fs';
import glob from 'glob';
import ejs from 'ejs';
import btoa from 'btoa';

import { formatBytes, getCommonPathPrefix } from './helpers';
import { ExploreResult, FileSizeMap } from './api';

export const UNMAPPED = '<unmapped>';

export function reportUnmappedBytes(data: ExploreResult): void {
  const unmappedBytes = data.files[UNMAPPED];
  if (unmappedBytes) {
    const totalBytes = data.totalBytes;
    const pct = (100 * unmappedBytes) / totalBytes;

    const bytesString = pct.toFixed(2);

    console.warn(
      `[${data.bundleName}] Unable to map ${unmappedBytes}/${totalBytes} bytes (${bytesString}%)`
    );
  }
}

/**
 * Create a combined result where each of the inputs is a separate node under the root.
 */
function makeMergedBundle(exploreResults: ExploreResult[]): ExploreResult {
  let totalBytes = 0;
  const files: Record<string, number> = {};

  // Remove any common prefix to keep the visualization as simple as possible.
  const commonPrefix = getCommonPathPrefix(exploreResults.map(r => r.bundleName));

  for (const result of exploreResults) {
    totalBytes += result.totalBytes;
    const prefix = result.bundleName.slice(commonPrefix.length);

    Object.keys(result.files).forEach(fileName => {
      const size = result.files[fileName];
      files[prefix + '/' + fileName] = size;
    });
  }

  return {
    bundleName: '[combined]',
    totalBytes,
    files,
  };
}

interface WebTreeMapNode {
  name: string;
  data: {
    $area: number;
  };
  children: WebTreeMapNode[];
}

/**
 * Covert file size map to webtreemap data
 */
function getWebTreeMapData(files: FileSizeMap): WebTreeMapNode {
  function newNode(name: string): WebTreeMapNode {
    return {
      name: name,
      data: {
        $area: 0,
      },
      children: [],
    };
  }

  const treeData = newNode('/');

  function addNode(path: string, size: number): void {
    const parts = path.split('/');
    let node = treeData;

    node.data['$area'] += size;

    parts.forEach(part => {
      let child = node.children.find(child => child.name === part);

      if (!child) {
        child = newNode(part);
        node.children.push(child);
      }

      node = child;
      node.data['$area'] += size;
    });
  }

  function addSizeToTitle(node: WebTreeMapNode, total: number): void {
    const size = node.data['$area'],
      pct = (100.0 * size) / total;

    node.name += ` • ${formatBytes(size)} • ${pct.toFixed(1)}%`;
    node.children.forEach(child => {
      addSizeToTitle(child, total);
    });
  }

  for (const source in files) {
    addNode(source, files[source]);
  }
  addSizeToTitle(treeData, treeData.data['$area']);

  return treeData;
}

/**
 *  Generate HTML file content for specified files
 */
export function generateHtml(exploreResults: ExploreResult[]): string {
  const assets = {
    webtreemapJs: btoa(fs.readFileSync(require.resolve('./vendor/webtreemap.js'))),
    webtreemapCss: btoa(fs.readFileSync(require.resolve('./vendor/webtreemap.css'))),
  };

  // Create a combined bundle if applicable
  if (exploreResults.length > 1) {
    exploreResults = [makeMergedBundle(exploreResults)].concat(exploreResults);
  }

  // Get bundles info to generate select
  const bundles = exploreResults.map(data => ({
    name: data.bundleName,
    size: formatBytes(data.totalBytes),
  }));

  // Get webtreemap data to update map on bundle select
  const treeDataMap = exploreResults.reduce<Record<string, WebTreeMapNode>>((result, data) => {
    result[data.bundleName] = getWebTreeMapData(data.files);

    return result;
  }, {});

  const template = fs.readFileSync(path.join(__dirname, 'tree-viz.ejs')).toString();

  return ejs.render(template, {
    bundles,
    treeDataMap,
    webtreemapJs: assets.webtreemapJs,
    webtreemapCss: assets.webtreemapCss,
  });
}

export interface Bundle {
  codePath: string;
  mapPath?: string;
}

/**
 * Expand codePath and mapPath into a list of { codePath, mapPath } pairs
 * @see https://github.com/danvk/source-map-explorer/issues/52
 * @param codePath Path to bundle file or glob matching bundle files
 * @param [mapPath] Path to bundle map file
 */
export function getBundles(codePath: string, mapPath?: string): Bundle[] {
  if (codePath && mapPath) {
    return [
      {
        codePath,
        mapPath,
      },
    ];
  }

  const filenames = glob.sync(codePath);
  const mapFilenames = glob.sync(codePath + '.map');

  return filenames
    .filter(filename => !filename.endsWith('.map'))
    .map<Bundle>(filename => ({
      codePath: filename,
      mapPath: mapFilenames.find(mapFilename => mapFilename === `${filename}.map`),
    }));
}
