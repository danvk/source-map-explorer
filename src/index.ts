import { explore } from './api';
import { UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY, NO_SOURCE_KEY } from './explore';

export { explore, UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY, NO_SOURCE_KEY };

export default explore;

// Export all interfaces from index.ts to avoid type exports in compiled js code. See https://github.com/babel/babel/issues/8361

export type FileSizeMap = Record<string, number>;

export interface FileSizes {
  files: FileSizeMap;
  unmappedBytes: number;
  eolBytes: number;
  sourceMapCommentBytes: number;
  totalBytes: number;
}

export type ErrorCode =
  | 'Unknown'
  | 'NoBundles'
  | 'NoSourceMap'
  | 'OneSourceSourceMap'
  | 'UnmappedBytes'
  | 'InvalidMappingLine'
  | 'InvalidMappingColumn'
  | 'CannotSaveFile'
  | 'CannotCreateTempFile'
  | 'CannotOpenTempFile';

export type File = string | Buffer;

export type ReplaceMap = Record<string, string>;

export type OutputFormat = 'json' | 'tsv' | 'html';

/** Represents single bundle */
export interface Bundle {
  code: File;
  map?: File;
}

export interface ExploreOptions {
  /** Exclude "unmapped" bytes from the output */
  onlyMapped?: boolean;
  /** Exclude source map comment size from output */
  excludeSourceMapComment?: boolean;
  /** Output result as a string */
  output?: {
    format: OutputFormat;
    /** Filename to save output to */
    filename?: string;
  };
  /** Disable removing prefix shared by all sources */
  noRoot?: boolean;
  /** Replace "this" by "that" map */
  replaceMap?: ReplaceMap;
}

export interface ExploreResult {
  bundles: ExploreBundleResult[];
  /** Result as a string - either JSON, TSV or HTML */
  output?: string;
  errors: ExploreErrorResult[];
}

export interface ExploreBundleResult extends FileSizes {
  bundleName: string;
}

export interface ExploreErrorResult {
  bundleName: string;
  code: string;
  message: string;
  error?: NodeJS.ErrnoException;
  isWarning?: boolean;
}

export type BundlesAndFileTokens = (Bundle | string)[] | Bundle | string;

// TODO: Remove when https://github.com/mozilla/source-map/pull/374 is merged
declare module 'source-map' {
  export interface MappingItem {
    lastGeneratedColumn: number | null;
  }
}
