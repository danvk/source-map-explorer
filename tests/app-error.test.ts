import snapshot from '@smpx/snap-shot-it';

import { getErrorMessage, ErrorContext } from '../src/app-error';

describe('app-error', () => {
  describe('getErrorMessage', () => {
    const tests = [
      { code: 'NoBundles' },
      { code: 'NoSourceMap' },
      { code: 'OneSourceSourceMap', filename: 'foo.min.js' },
      { code: 'UnmappedBytes', totalBytes: 100, unmappedBytes: 70 },
      { code: 'InvalidMappingLine', generatedLine: 60, maxLine: 57 },
      {
        code: 'InvalidMappingColumn',
        generatedLine: 60,
        generatedColumn: 81,
        maxColumn: 80,
      },
      { code: 'CannotSaveFile' },
      { code: 'CannotCreateTempFile' },
      {
        code: 'CannotOpenTempFile',
        error: Buffer.from('The system cannot find the file ?C:\\foo.htm'),
        tempFile: 'C:\\foo.htm',
      },
      { code: 'UnknownCode' },
    ];

    tests.forEach(context => {
      it(`should create message for '${context.code}'`, () => {
        snapshot(getErrorMessage(context as ErrorContext));
      });
    });
  });
});
