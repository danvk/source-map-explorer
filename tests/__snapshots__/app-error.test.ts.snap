exports['app-error getErrorMessage should create message for \'CannotCreateTempFile\' 1'] = `
Unable to create a temporary HTML file
`

exports['app-error getErrorMessage should create message for \'CannotOpenCoverageFile\' 1'] = `
Unable to open/parse coverage file
`

exports['app-error getErrorMessage should create message for \'CannotOpenTempFile\' 1'] = `
Unable to open web browser. The system cannot find the file ?C:\\foo.htm
Either run with --html, --json, --tsv, --file, or view HTML for the visualization at:
C:\\foo.htm
`

exports['app-error getErrorMessage should create message for \'CannotSaveFile\' 1'] = `
Unable to save HTML to file
`

exports['app-error getErrorMessage should create message for \'InvalidMappingColumn\' 1'] = `
Your source map refers to generated column 81 on line 60, but the source only contains 80 column(s) on that line.
Check that you are using the correct source map.
`

exports['app-error getErrorMessage should create message for \'InvalidMappingLine\' 1'] = `
Your source map refers to generated line 60, but the source only contains 57 line(s).
Check that you are using the correct source map.
`

exports['app-error getErrorMessage should create message for \'NoBundles\' 1'] = `
No file(s) provided
`

exports['app-error getErrorMessage should create message for \'NoCoverageMatches\' 1'] = `
No matched bundles found for coverages
`

exports['app-error getErrorMessage should create message for \'NoSourceMap\' 1'] = `
Unable to find a source map.
See https://github.com/danvk/source-map-explorer/blob/master/README.md#generating-source-maps
`

exports['app-error getErrorMessage should create message for \'OneSourceSourceMap\' 1'] = `
Your source map only contains one source (foo.min.js).
This can happen if you use browserify+uglifyjs, for example, and don't set the --in-source-map flag to uglify.
See https://github.com/danvk/source-map-explorer/blob/master/README.md#generating-source-maps
`

exports['app-error getErrorMessage should create message for \'UnknownCode\' 1'] = `
Unknown error
`

exports['app-error getErrorMessage should create message for \'UnmappedBytes\' 1'] = `
Unable to map 70/100 bytes (70.00%)
`
